client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx

```diff
@@ -0,0 +1,94 @@
+import { FEATURE_SET_PRIMARY_CUSTOM_DOMAIN } from '@automattic/calypso-products';
+import { useTranslate } from 'i18n-calypso';
+import { useState, ChangeEvent, useEffect } from 'react';
+import FormSelect from 'calypso/components/forms/form-select';
+import type { DomainData, SiteDetails } from '@automattic/data-stores';
+
+import './style.scss';
+
+type PrimaryDomainSelectorProps = {
+	domains: undefined | DomainData[];
+	userCanSetPrimaryDomains: boolean;
+	site: undefined | null | SiteDetails;
+	onSetPrimaryDomain: ( domain: string, onComplete: () => void ) => void;
+};
+
+const PrimaryDomainSelector = ( {
+	domains,
+	site,
+	userCanSetPrimaryDomains,
+	onSetPrimaryDomain,
+}: PrimaryDomainSelectorProps ) => {
+	const [ selectedDomain, setSelectedDomain ] = useState< undefined | string >( undefined );
+	const [ isSettingPrimaryDomain, setIsSettingPrimaryDomain ] = useState< boolean >( false );
+
+	const translate = useTranslate();
+
+	useEffect( () => {
+		if ( domains?.length ) {
+			const primaryDomain = domains.find( ( domain ) => {
+				return domain.primary_domain;
+			} );
+			if ( primaryDomain ) {
+				setSelectedDomain( primaryDomain.domain );
+			}
+		}
+	}, [ domains ] );
+
+	if ( ! domains || ! site ) {
+		return null;
+	}
+
+	const isOnFreePlan = site?.plan?.is_free ?? false;
+
+	const shouldUpgradeToMakeDomainPrimary = ( domain: DomainData ): boolean => {
+		return (
+			! userCanSetPrimaryDomains &&
+			isOnFreePlan &&
+			( domain.type === 'registered' || domain.type === 'mapping' ) &&
+			! domain.current_user_can_create_site_from_domain_only &&
+			! domain.wpcom_domain &&
+			! domain.is_wpcom_staging_domain &&
+			( site?.plan?.features.active.includes( FEATURE_SET_PRIMARY_CUSTOM_DOMAIN ) ?? false )
+		);
+	};
+
+	const validPrimaryDomains = domains.filter( ( domain ) => {
```

Comment:
Would it make sense to filter this out to only custom domains using `isCustomDomain`? It seems counter-intuitive to encourage users to set their primary domain as one which isn't a custom domain.

Reply:
See my comment https://github.com/Automattic/wp-calypso/pull/88383#issuecomment-1994990432